src/cargo/core/compiler/future_incompat.rs

```diff
@@ -24,13 +52,239 @@ pub struct Diagnostic {
+        if on_disk_reports.version != 0 {
+            bail!("unable to read reports; reports were saved from a future version of Cargo");
+        }
+        Ok(on_disk_reports)
+    }
+
+    /// Returns the most recent report ID.
+    pub fn last_id(&self) -> u32 {
+        self.reports.last().map(|r| r.id).unwrap()
+    }
+
+    pub fn get_report(&self, id: u32, config: &Config) -> CargoResult<String> {
+        let report = self.reports.iter().find(|r| r.id == id).ok_or_else(|| {
+            let available = iter_join(self.reports.iter().map(|r| r.id.to_string()), ", ");
+            format_err!(
+                "could not find report with ID {}\n\
+                 Available IDs are: {}",
+                id,
+                available
+            )
+        })?;
+        let report = if config.shell().err_supports_color() {
+            report.report.clone()
+        } else {
+            strip_ansi_escapes::strip(&report.report)
+                .map(|v| String::from_utf8(v).expect("utf8"))
+                .expect("strip should never fail")
+        };
+        Ok(report)
+    }
+}
+
+fn render_report(
+    ws: &Workspace<'_>,
+    per_package_reports: &[FutureIncompatReportPackage],
+) -> String {
+    let mut per_package_reports: Vec<_> = per_package_reports.iter().collect();
+    per_package_reports.sort_by_key(|r| r.package_id);
+    let mut rendered = String::new();
+    for per_package in &per_package_reports {
+        rendered.push_str(&format!(
+            "The package `{}` currently triggers the following future \
+             incompatibility lints:\n",
+            per_package.package_id
+        ));
+        for item in &per_package.items {
+            rendered.extend(
+                item.diagnostic
+                    .rendered
+                    .lines()
+                    .map(|l| format!("> {}\n", l)),
+            );
+        }
+        rendered.push('\n');
+    }
+    if let Some(s) = render_suggestions(ws, &per_package_reports) {
+        rendered.push_str(&s);
+    }
+    rendered
+}
+
+fn render_suggestions(
+    ws: &Workspace<'_>,
+    per_package_reports: &[&FutureIncompatReportPackage],
+) -> Option<String> {
+    // This in general ignores all errors since this is opportunistic.
+    let _lock = ws.config().acquire_package_cache_lock().ok()?;
+    // Create a set of updated registry sources.
+    let map = SourceConfigMap::new(ws.config()).ok()?;
+    let package_ids: BTreeSet<_> = per_package_reports
+        .iter()
+        .map(|r| r.package_id)
+        .filter(|pkg_id| pkg_id.source_id().is_registry())
+        .collect();
+    let source_ids: HashSet<_> = package_ids
+        .iter()
+        .map(|pkg_id| pkg_id.source_id())
+        .collect();
+    let mut sources: HashMap<_, _> = source_ids
+        .into_iter()
+        .filter_map(|sid| {
+            let unlocked = sid.clone().with_precise(None);
+            let mut source = map.load(unlocked, &HashSet::new()).ok()?;
```

Comment:
This may have been preexisting, but this looks a little worrisome looking at this again. Does this means that, on faiulre, we might try to re-update the registry?

(I'd expect here that we'd just use w/e state the registry had without updating anything)

Reply:
Yea, I waffled a little bit on this.  My thinking is that these future-incompat errors should be very rare. Thinking about it more, this is a bad idea to hit the network on every build.  My only thought is to maybe only update if the index is sufficiently outdated (once every few days?). 

I changed it so that it won't update.